pkg/services/datasources/service/store.go

```diff
@@ -77,6 +73,28 @@ func (ss *SqlStore) getDataSource(ctx context.Context, query *datasources.GetDat
 	return datasource, nil
 }
 
+func (ss *SqlStore) getDataSource(ctx context.Context, query *datasources.GetDataSourceQuery, sess *db.Session) (*datasources.DataSource, error) {
+	if query.OrgID == 0 || (query.ID == 0 && len(query.Name) == 0 && len(query.UID) == 0) {
+		return nil, datasources.ErrDataSourceIdentifierNotSet
+	}
+
+	if err := util.ValidateUID(query.UID); len(query.UID) > 0 && err != nil {
```

Comment:
I wonder if we should optimize for a valid UID and query with the provided UID first. You even mention below "I assume there won't many". Then if no match, validate, fix and re-query using the fixed uid?

After consideration, I see one potential (security) flaw with this approach and that is that this could be used for someone to craft a datasource with a uid to get access to a datasource they don't have access to which is not good. To secure this we probably need to implement this higher up where permission checks happens 😢

Reply:
> I wonder if we should optimize for a valid UID and query with the provided UID first. You even mention below "I assume there won't many". Then if no match, validate, fix and re-query using the fixed uid?

Good point.

> After consideration, I see one potential (security) flaw with this approach and that is that this could be used for someone to craft a datasource with a uid to get access to a datasource they don't have access to which is not good. To secure this we probably need to implement this higher up where permission checks happens 😢

True. It's a bit of a corner case because the user would need to have specific access for the wrong UID (e.g. `custom/uid`) but not for the valid representation (`custom-uid`) plus it would require that the datasource with the wrong UID doesn't exist (so we fallback to check the valid UID as per your comment above).

In any case, we may not need this if the datasources with wrong UIDs are not used (since they are just pre-created). I am going to put this behind a feature flag, keeping the logging so we can get more information.

Reply:
>  it would require that the datasource with the wrong UID doesn't exist 

Yes. So even if we did your original validation/fix or my proposal they both has the same flaw

> In any case, we may not need this if the datasources with wrong UIDs are not used (since they are just pre-created)

Think you can expand on this, pre-created especially?

Reply:
> Think you can expand on this, pre-created especially?

I just meant automatically provisioned, without user interaction.